lib/modules/manager/npm/extract/yarnrc.ts

```diff
@@ -0,0 +1,46 @@
+import { load } from 'js-yaml';
+import { z } from 'zod';
+import { logger } from '../../../../logger';
+
+const YarnrcYmlSchema = z.object({
+  npmRegistryServer: z.string().optional(),
+  npmScopes: z
+    .record(
+      z.object({
+        npmRegistryServer: z.string().optional(),
+      })
+    )
+    .optional(),
+});
+
+export type YarnConfig = z.infer<typeof YarnrcYmlSchema>;
+
+export function loadConfigFromYarnrcYml(yarnrcYml: string): YarnConfig | null {
+  try {
+    return YarnrcYmlSchema.parse(
+      load(yarnrcYml, {
+        json: true,
+      })
+    );
+  } catch (err) {
+    logger.warn({ yarnrcYml, err }, `Failed to load yarnrc file`);
+    return null;
+  }
+}
+
+export function resolveRegistryUrl(
+  packageName: string,
+  yarnConfig: YarnConfig
+): string | null {
+  if (yarnConfig.npmScopes) {
+    for (const scope in yarnConfig.npmScopes) {
+      if (packageName.startsWith(`@${scope}`)) {
```

Comment:
```suggestion
      if (packageName.startsWith(`@${scope}/`)) {
```
otherwise it can wrongly match scopes

Reply:
Good catch. A test case with

```yml
npmScopes:
  foo:
    npmRegistryServer: https://npm-foo.example.com
```

and `resolveRegistryUrl('@foo-bar/a-package');` will prevent a regression too.

Reply:
Ah, thanks for the sharp look. Added the missing test case and fixed the issue.